### PR TITLE
Speed up runtime deobfuscation

### PR DESCRIPTION
**Pull request by ichttt**
This PR aims to improve startup time, esp on large modpacks. This is done by changing the deobf transformer to not expand frames, which, as the javadoc says, is slow. This works for a simple renaming transformer as well, but we need to change it to a ´ClassRemapper´ instead of a ´RemappingClassAdapter´, as the ´RemappingClassAdapter´ class requires ´EXPAND_FRAMES´.
See https://gitlab.ow2.org/asm/asm/issues/317576 for some background information.

### CODE REVIEW

File: src/main/java/net/minecraftforge/fml/common/asm/transformers/DeobfuscationTransformer.java [hunk 1]
```diff
@@ -21,6 +21,7 @@
 
 import java.util.Arrays;
 
+import com.google.common.base.Stopwatch;
```
Comment (reece2d) [inline src/main/java/net/minecraftforge/fml/common/asm/transformers/DeobfuscationTransformer.java]: Remember to remove unnecessary dependancies

**Review by reece2d — CHANGES_REQUESTED** [review]

**Review by tterrag1098 — APPROVED** [review]
This looks okay to me, but I'll leave it to @LexManos to give the final review since he knows more about this system.

**Comment by ichttt** [comment]
I also added check to the SideTransformer, so classes that didn't changed don't have their class bytes recalculated. The same check is already present in all other transformers.

**Review by mezz — APPROVED** [review]
I was curious why `ClassReader.EXPAND_FRAMES` was needed before but didn't find much information on it.
There are some [pretty old bugs](https://jira.terracotta.org/jira/si/jira.issueviews:issue-html/CDV-286/CDV-286.html) about it being required for some specific `classReader.accept` implementations but I haven't seen anything recent so I think your analysis is the right one.

If anyone else is curious, the original `DeobfuscationTransformer` was written here: https://github.com/MinecraftForge/FML/commit/b5d5c95b3364075e111afe79395f23198426645c#diff-bb8c8be01d90f6d6dde226c050f76b44

**Comment by ichttt** [comment]
You are right about these old bugs. This is because The RemappingMethodAdapter is a subclass of LocalVariablesSorter, which requires EXPAND_FRAMES to sort. However, variable sorting is not needed for a simple renaming transformer (see the linked issue in my original explenation), so the ASM authors added a new MethodRemapper, which doesn't subclass LocalVariableSorter. In theory this is a breaking change, but as the asm transformer has no use outside of asm context and coremods might break on any change, this shouldn't be a problem.

**Comment by Darkhax** [comment]
I think this is a good change but it would be nice to have some benchmarks that go along with it, even if it's just a simple one.

**Comment by ichttt** [comment]
I did some testing with obfuscated minecraft with a modpack consisting of 171 mods and got the following results with [my profiler](https://gist.github.com/ichttt/88b650ead95b5415c7ee3972cfa123d9):
1.12.x with profiler patches:
SideTransformer: Total 1182 invocations 28400 mean 0.04163472
DeobfTransformer: Total 11174 invocations 28200 mean 0.39627618
My branch with profiler patches:
SideTransformer: Total 19 invocations 28400 mean 0.00068838027 (ms)
DeobfTransformer: Total 10060 invocations 28200 mean 0.3567684 (ms)

As the SideTransformer has close to no work on the client, the difference is most visible here, and the difference in the time needed for deobf is also visible.
I tested each version twice and got more or less the same results

These changes are not huge, but you have to keep in mind that classloading can also happen while playing. To smooth out the experience as much as possible, lower transformation times contribute to less lag spikes.

**Comment by Darkhax** [comment]
Thanks, you may want to show the actual decimal value for the second run so it's easier to compare. (0.00068838027)

**Comment by tterrag1098** [comment]
This is a far-reaching change with very little actual improvement. I'm not sure it's worth risking the many edge cases.

**Comment by ichttt** [comment]
Fair enough

**Pull request closed without merging**